### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Joeper214**
setting the zabbix templates dynamically. and getting the linked/selected templates of the current host_id.

### CODE REVIEW

**Comment by pocke** [comment]
### TODOs

- [ ] Test is broken
- [ ] Code format
- [ ] Document comment by YARD
- [ ] Zabbix API parameters

Let's do tomorrow!

**Comment by Joeper214** [comment]
Yes! I'm looking forward. thanks. 

お疲れ様でした!

**Comment by pocke** [comment]
Code format
--------------

We use the EditorConifg(http://editorconfig.org/).
By using the EditorConfig, we can to unify the format of code(e.g. indent, encoding...).
Can you install the plug-in of EditorConfig to your editor?

**Comment by pocke** [comment]
Test is broken
-------------

I fixed it. But it's still broken on ruby-head.... 
I will make corrections Continuously.

**Comment by pocke** [comment]
Document comment by YARD
----------------------------

We use the YARD(http://yardoc.org/).
YARD is a Ruby documentation tool.

Please write a short description and type of method parameter and return value on the method definition.
(e.g. https://github.com/skyarch-networks/skyhopper/blob/develop/app/models/zabbix.rb#L177-L179)

File: app/models/zabbix.rb [hunk 1]
```diff
@@ -92,6 +55,22 @@ def templates_link_host(physical_id, template_names)
     )
   end
 
+  #get available tempaltes from zabbix
+  def available_templates()
+    templates = @sky_zabbix.template.get(output: "extend", filter: {name: ""}).map{|x|x['name']}
```
Comment (pocke) [inline app/models/zabbix.rb]: `filter: {name: ""}` is unnecessary.


When Parameter of output is `extend`, Zabbix return value contains all value.
But we want to use merely `name`.
So, this parameter is only enough `['name']`.

File: app/controllers/monitorings_controller.rb [hunk 2]
```diff
@@ -33,12 +34,18 @@ def show
       @before_register = true
 
       #get/load available zabbix templates set to static first.
-      @templates = Zabbix::AvailableTemplates.map{|t| {name: t, checked: false}}
+      @templates = @zabbix.available_templates().map{|t| {name: t, checked: false}}
```
Comment (pocke) [inline app/controllers/monitorings_controller.rb]: We like to omit parentheses for method calls without any arguments.
Definition of method without any arguments is the same(e.g. https://github.com/skyarch-networks/skyhopper/pull/15/files#diff-61e8618dd4e0593bc4c0edf0c085c256R59).

**Comment by pocke** [comment]
Test is broken
-------------

> I fixed it. But it's still broken on ruby-head.... 
> I will make corrections Continuously.

FIxed! https://travis-ci.org/skyarch-networks/skyhopper/builds/76040270

File: app/models/zabbix.rb (hunk 1 above)
Comment (Joeper214) [inline app/models/zabbix.rb]: Okay I will install editor config first before fixing this. thanks.

File: app/controllers/monitorings_controller.rb (hunk 2 above)
Comment (Joeper214) [inline app/controllers/monitorings_controller.rb]: so the function call should be like this? @zabbix.available_templates.map()?

Comment (Joeper214) [inline app/controllers/monitorings_controller.rb]: Ahh. I see. the ruby way of defining a methods does not need parentheses. thanks. 

I saw it in this documentation. http://ruby-doc.com/docs/ProgrammingRuby/html/tut_methods.html

**Comment by Joeper214** [comment]
I am having trouble in installing editorconfig-core-c.

https://github.com/editorconfig/editorconfig-core-c/blob/master/INSTALL.md

I think I need help on this. :)

File: app/controllers/monitorings_controller.rb (hunk 2 above)
Comment (pocke) [inline app/controllers/monitorings_controller.rb]: Yes, that's right!

From `@zabbix.available_template()` to `@zabbix.abailable_template`
From `def available_template()` to `def available_template`

**Comment by pocke** [comment]
Do you use the Atom editor?
If you use the Atom, there is maybe no need to install editorconfig-core-c.
You need to install only [atom-editorconifg](https://github.com/sindresorhus/atom-editorconfig).

File: app/controllers/monitorings_controller.rb [hunk 3]
```diff
@@ -33,12 +34,18 @@ def show
...
       return
     end
 
     @monitor_selected_common   = @infra.master_monitorings.where(is_common: true)
     @monitor_selected_uncommon = @infra.master_monitorings.where(is_common: false)
+
+    resources = @infra.resources.ec2
+    resources.each do |resource|
+      @linked_templates = @zabbix.get_linked_templates(resource.physical_id).map{|t| {name: t}}
+    end
```
Comment (pocke) [inline app/controllers/monitorings_controller.rb]: When resources is plural, `@linked_templates` has templates of last resource.
Your code will work in the same way with the following code.

```ruby
@linked_templates = @zabbix.get_linked_templates(resources.last.physical_id).map{|t| {name: t}}
```
Can you change to include templates of all resource into `@linked_templates` ?

**Comment by Joeper214** [comment]
No. I changed to Emacs editor. in the Ec2 instance.

File: app/controllers/monitorings_controller.rb (hunk 3 above)
Comment (Joeper214) [inline app/controllers/monitorings_controller.rb]: sorry. this line of code ist still incomplete. I haven't push the working version yet.

Comment (Joeper214) [inline app/controllers/monitorings_controller.rb]: cmake

**Comment by pocke** [comment]
Could you try to `yum install pcre-devel`?

**Comment by Joeper214** [comment]
It works!. Thank you very much.

**Comment by pocke** [comment]
You're welcome! :cat:

**Comment by Joeper214** [comment]
Anyway, How can we check/test if the editorconfig is working? I already installed it.

**Comment by pocke** [comment]
Editroconfig-emacs supports trim_trailing_whitespace.
Please open a file of SkyHopper.
Then, input a space at the end of the line, and save.
If if disappears blank that you entered, plug-in installation is successful!

File: app/views/monitorings/_modal_edit_templates.html.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+        <h4 class="modal-title"><%= t 'monitoring.before_reg_title' %></h4>
+      </div>
+      <div class="modal-body">
+
+        <div class="col-md-12 col-sm-12">
+          <form>
+            <div class="checkbox" v-repeat="t: templates">
+              <label class="checkbox">
+                <input type="checkbox" v-attr="checked: t.checked" v-model="t.checked" id="inputEmail3">{{t.name}}
```
Comment (pocke) [inline app/views/monitorings/_modal_edit_templates.html.erb]: Does this `id=inputEmail3` is not necessary?

File: app/assets/javascripts/infrastructures.js [hunk 5]
```diff
@@ -299,6 +299,21 @@
           })(t('monitoring.msg.created'));
         }).fail(alert_and_show_infra);
       },
+      edit_templates: function () {
+        if (!this.has_selected) {return;}
+
```
Comment (pocke) [inline app/assets/javascripts/infrastructures.js]: Can you display the message of loading?

**Comment by pocke** [comment]
If the Update template is repeated several times, I may receive an error message such as the following.
default

**Comment by Joeper214** [comment]
yes. because it means that you cannot add duplicate machine host like windows, mac os and linux. so zabbix returns an error.

File: app/assets/javascripts/infrastructures.js (hunk 5 above)
Comment (Joeper214) [inline app/assets/javascripts/infrastructures.js]: okay.

**Comment by pocke** [comment]
@Joeper214  Thanks!
Can you display a descriptive error message?

**Comment by Joeper214** [comment]
I don't know how to do that. because zabbix API normally returns this error message.

zabbix_error

**Comment by pocke** [comment]
Hmm... if we find a solution, let's solve this!

**Comment by Joeper214** [comment]
I think we can just change the error message that is understandable to the users?

**Comment by pocke** [comment]
I think we can change this.
For example, if the error message has `"agent.hostname" already linked to host`, modal displays "Can not include more than a 'Template OS *.'" to users.
Or if user choose a plurality of "Template OS *", modal displays that.

I think the latter solution is better.
Because, the error message may vary by update of Zabbix.

File: app/controllers/monitorings_controller.rb [hunk 6]
```diff
@@ -33,16 +34,54 @@ def show
...
+
+    @templates = merged
+
     @resources = @infra.resources.ec2
 
   end
 
+  # POST /monitorings/:id/edit_templates
+  def edit_templates
```
Comment (pocke) [inline app/controllers/monitorings_controller.rb]: `update_template` is suitable for name of this method than `edit_template`.

File: app/controllers/monitorings_controller.rb [hunk 7]
```diff
@@ -33,16 +34,54 @@ def show
...
+    # compare if the previous templates was removed and push to clear list
+    prev_templates.each do |prev|
+      if new_templates.include?(prev)
+        # new_templates.pop(prev)
+      else
+        clear_templates.push(prev)
+      end
+    end
+
+    @zabbix.templates_update_host(resources.last.physical_id, new_templates, clear_templates)
```
Comment (pocke) [inline app/controllers/monitorings_controller.rb]: Please update all resources.

**Comment by pocke** [comment]
Editing of template is a setting.
But there isn't this feature on `edit-monitoring` tabpane.
I want to use this feature in both of the tabpane.

File: app/controllers/monitorings_controller.rb (hunk 6 above)
Comment (Joeper214) [inline app/controllers/monitorings_controller.rb]: okay. sorry for the confusion. I'll change this.

File: app/controllers/monitorings_controller.rb (hunk 7 above)
Comment (Joeper214) [inline app/controllers/monitorings_controller.rb]: :+1:

**Comment by Joeper214** [comment]
okay. so, do you mean that the tab pane is the selection or the features for the zabbix/monitoring configuration?

File: app/controllers/monitorings_controller.rb (hunk 7 above)
Comment (pocke) [inline app/controllers/monitorings_controller.rb]: :+1: :+1:

**Comment by pocke** [comment]
Yes, thats right!

**Comment by Joeper214** [comment]
Okay. I work on that after I configured my unison. thanks!

**Comment by Joeper214** [comment]
I was able to implement the fix. please check the dev. branch. thanks!

**Comment by pocke** [comment]
Sorry, I'll check the next Tuesday.

**Comment by Joeper214** [comment]
It's okay. Take your time. Thank you very much!.

**Comment by pocke** [comment]
One pull request should have one feature only.
This is because, there is a reason such as the following.

- Reviews becomes complicated.
- If a feature isn't good, I can't merge all features.
- It takes a lot time until the merge.

Could you divide pull request by feature from next time?

**Comment by Joeper214** [comment]
Yeah sure. Next time.

**Comment by pocke** [comment]
Thanks!

> Editing of template is a setting.
> But there isn't this feature on edit-monitoring tabpane.
> I want to use this feature in both of the tabpane.

I have finished reviewing the above.
It good!

**Comment by Joeper214** [comment]
Yeah thanks. But the test still failed?

**Comment by pocke** [comment]
Yes, Test can't find a celluloid gem.

> Could not find celluloid-0.16.1 in any of the sources
https://travis-ci.org/skyarch-networks/skyhopper/jobs/77103273

I try to fix it.

**Comment by pocke** [comment]
I fixed this problem on ab2e9c516e645a52f5bd14fca376ebbdff6696f6.
Please merge `develop` branch into `dev` branch.

**Comment by Joeper214** [comment]
Done merging. Thanks.

**Comment by pocke** [comment]
This PR has some task.
But I want to merge PR.
Because, if I don't merge, This PR will become very huge.


If there is no objection, I will merge this.
And I'll create new task on backlog.

**Comment by Joeper214** [comment]
Okay. Please do.

**Comment by pocke** [comment]
I merged it! thanks :+1: 

I wrote task in backlog.
Please create new branch from develop.

**Comment by Joeper214** [comment]
alright!.  thanks. :+1:

**Pull request merged**